File: app/src/main/java/com/itachi1706/cheesecakeutilities/Modules/MSLIntegration/model/CalendarModel.java
package com.itachi1706.cheesecakeutilities.Modules.MSLIntegration.model;

import com.google.api.services.calendar.model.Calendar;
import com.google.api.services.calendar.model.CalendarListEntry;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by Kenneth on 13/1/2019.
 * for com.itachi1706.cheesecakeutilities.Modules.MSLIntegration in CheesecakeUtilities
 */
public class CalendarModel {

    private final Map<String, CalendarInfo> calendars = new HashMap<String, CalendarInfo>();

    public int size() {
        synchronized (calendars) {
            return calendars.size();
        }
    }

    public void remove(String id) {
        synchronized (calendars) {
            calendars.remove(id);
        }
    }

    public CalendarInfo get(String id) {
        synchronized (calendars) {
            return calendars.get(id);
        }
    }

    public void add(Calendar calendarToAdd) {
        synchronized (calendars) {
            CalendarInfo found = get(calendarToAdd.getId());
            if (found == null) {
                calendars.put(calendarToAdd.getId(), new CalendarInfo(calendarToAdd));
            } else {
                found.update(calendarToAdd);
            }
        }
    }

    public void add(CalendarListEntry calendarToAdd) {
        synchronized (calendars) {
            CalendarInfo found = get(calendarToAdd.getId());
            if (found == null) {
                calendars.put(calendarToAdd.getId(), new CalendarInfo(calendarToAdd));
            } else {
                found.update(calendarToAdd);
            }
        }
    }

    public void reset(List<CalendarListEntry> calendarsToAdd) {
        synchronized (calendars) {
            calendars.clear();
            for (CalendarListEntry calendarToAdd : calendarsToAdd) {
                add(calendarToAdd);
            }
        }
    }

    public CalendarInfo[] toSortedArray() {
        synchronized (calendars) {
            List<CalendarInfo> result = new ArrayList<CalendarInfo>();
            for (CalendarInfo calendar : calendars.values()) {
                result.add(calendar.clone());
            }
            Collections.sort(result);
            return result.toArray(new CalendarInfo[0]);
        }
    }
}

File: app/src/main/java/com/itachi1706/cheesecakeutilities/Modules/MSLIntegration/model/CalendarModel.kt
package com.itachi1706.cheesecakeutilities.Modules.MSLIntegration.model

import com.google.api.services.calendar.model.Calendar
import com.google.api.services.calendar.model.CalendarListEntry
import java.util.*

/**
 * Created by Kenneth on 13/1/2019.
 * for com.itachi1706.cheesecakeutilities.Modules.MSLIntegration in CheesecakeUtilities
 */
class CalendarModel {

    private val calendars = HashMap<String, CalendarInfo>()

    fun size(): Int {
        synchronized(calendars) {
            return calendars.size
        }
    }

    fun remove(id: String) {
        synchronized(calendars) {
            calendars.remove(id)
        }
    }

    operator fun get(id: String): CalendarInfo? {
        synchronized(calendars) {
            return calendars[id]
        }
    }

    fun add(calendarToAdd: Calendar) {
        synchronized(calendars) {
            val found = get(calendarToAdd.id)
            found?.update(calendarToAdd)
                    ?: calendars.put(calendarToAdd.id, CalendarInfo(calendarToAdd))
        }
    }

    fun add(calendarToAdd: CalendarListEntry) {
        synchronized(calendars) {
            val found = get(calendarToAdd.id)
            found?.update(calendarToAdd)
                    ?: calendars.put(calendarToAdd.id, CalendarInfo(calendarToAdd))
        }
    }

    fun reset(calendarsToAdd: List<CalendarListEntry>) {
        synchronized(calendars) {
            calendars.clear()
            for (calendarToAdd in calendarsToAdd) {
                add(calendarToAdd)
            }
        }
    }

    fun toSortedArray(): Array<CalendarInfo> {
        synchronized(calendars) {
            val result = ArrayList<CalendarInfo>()
            for (calendar in calendars.values) {
                result.add(calendar.clone())
            }
            result.sort()
            return result.toTypedArray()
        }
    }
}